add a functionalize() transform

bdhirsh opened the pull request:
Adds a functionalize() transform. You can use it like this:
```
def f(x):
    ....
out = vmap(functionalize(g))(batched_input)
```

There are some known rough edges (and probably some unknown ones too):
- logic with wrapped_tensor_scalars breaks (e.g. `torch.add(x, 1)`). This should probably be fixed sooner, because it could also affect conditional functionalization (if there happen to be composite ops that we need to functionalize, that use wrapped tensor scalars)
- nested functionalize() calls don't work yet

The functionalize() pass uses the same core machinery that's tested by https://github.com/pytorch/pytorch/pull/66101. It also has some extra logic to modify any mutated inputs after the pass completes, so I tested manually with this script:
```
import torch
from functorch import make_fx, grad, vmap, functionalize

def h2(x):
    tmp = torch.ones(2, 2)
    y = x.view(8)
    z0 = y.reshape(2, 4)
    z1 = z0.transpose(1, 0)
    z1.unsqueeze_(0)
    z1.squeeze_()
    z2, z3 = z1.split(2)
    z2.add_(tmp)
    return x

def h1(x):
    tmp = torch.ones(4)
    x.transpose_(1, 0)
    y = x[0]
    y.add_(tmp)
    return x

def g(x):
    tmp = torch.ones(2)
    y1, y2 = x.split(2)
    y1_view = y1.diagonal()
    y1_view.add_(tmp)
    return x

def f(x):
    tmp = torch.ones(2)
    y = x.permute(1, 0)
    z1 = y[0]
    z2 = z1.reshape(2, 2)
    z3 = z2.diagonal()
    z3.add_(tmp)
    return x



input_ = torch.zeros(4, 2)
ref_out1 = h2(input_)

input_ = torch.zeros(4, 2)
ref_out2 = h1(input_)

input_ = torch.zeros(4, 2)
ref_out3 = g(input_)

input_ = torch.zeros(4, 2)
ref_out4 = f(input_)

batched_input = torch.zeros(2, 4, 2)
out1 = vmap(functionalize(h2))(batched_input)

batched_input = torch.zeros(2, 4, 2)
out2 = vmap(functionalize(h1))(batched_input)

batched_input = torch.zeros(2, 4, 2)
print("BEFORE: " + str(batched_input))
out3 = vmap(functionalize(g))(batched_input)
print("AFTER: " + str(batched_input))

batched_input = torch.zeros(2, 4, 2)
print("BEFORE: " + str(batched_input))
out4 = vmap(functionalize(f))(batched_input)
print("AFTER: " + str(batched_input))

print(torch.allclose(out1[0], ref_out1))
print(torch.allclose(out2[0], ref_out2))
print(torch.allclose(out3[0], ref_out3))
print(torch.allclose(out4[0], ref_out4))
```

bdhirsh commented:
I think this is ready for another round. Latest changes:
- I rebased and fixed merged conflicts in DynamicLayer.cpp
- moved functionalize() to under `functorch.experimental`
- added a basic set of tests in `test_eager_transforms.py`

bdhirsh commented:
Nvm, looks like there's some interaction that broke the conditional functionalization tests. I'll take a look first.

bdhirsh commented:
Ok, I think this is fixed now. The bug: it looks like `DynamicLayer.cpp` adds all inactive "layer keys" to the TLS exclude set by default. Since my PR adds `DispatchKey::Functionalize` as a layer key, then `Functionalize` now gets added to the TLS exclude set whenever we hit a vmap layer.

The fix: The conditional functionalization fallback used to just add `Functionalize` to the TLS include set, but now it also explicitly removes `Functionalize` from the exclude set.

You have a handy RAII guard for doing that in `DynamicLayer.cpp`, so I moved it into a header file so I could re-use it in the fallback (I also left a comment that it might be useful to add in core at some point).

bdhirsh commented:
bleh, just realized I was using python 3.6 locally which for some reason passes all tests, but the remaining failures are with 3.7+. Taking a look at the new failures now.

bdhirsh commented:
Actually, it looks like the all of the failures are on master too (some recent pytorch core commit probably broke functorch and CI hasn't re-run yet)

zou3519 commented on test/test_eager_transforms.py:
```diff
@@ -2134,6 +2134,56 @@ def compute_loss(weights, image, target):
+        expected_outputs = f(inpt1)
+        actual_outputs = vmap(functionalize(f))(inpt2.unsqueeze(0))[0].squeeze()
+        # Check that outputs are the same
+        self.assertEqual(actual_outputs, expected_outputs)
```
Lol functionalize(f) could just return f and these tests wouldn't know better :D.

This seems fine for now though -- we'll probably find some better way of testing this in the future.

bdhirsh replied:
Yep and agreed :D

Long term, I think a combination of OpInfo tests in core + loggingtensor expecttests in functorch would be a lot better.

zou3519 commented on functorch/csrc/DynamicLayer.cpp:
```diff
@@ -312,6 +302,19 @@ std::ostream& operator<< (std::ostream& os, const std::vector<DynamicLayer>& dls
   return os;
 }
 
+void foreachTensorInplace(std::vector<IValue>& args, int64_t begin, int64_t end,
+    std::function<Tensor(const Tensor&)> func) {
+    foreachTensorInplaceWithIdx(args, begin, end, [&func](const Tensor& t, size_t) { return func(t); });
+}
```
Nice reuse!

zou3519 commented on functorch/csrc/DynamicLayer.cpp:
```diff
@@ -254,8 +244,8 @@ static Tensor unwrapIfDead(const Tensor& tensor) {
 
-void foreachTensorInplace(std::vector<IValue>& args, int64_t begin, int64_t end,
-    std::function<Tensor(const Tensor&)> func) {
+void foreachTensorInplaceWithIdx(std::vector<IValue>& args, int64_t begin, int64_t end,
```
nit: quick comment that this calls the function with increasing idx would be good, I had to confirm by reading the code

zou3519 replied:
Also, is this used anywhere?

bdhirsh replied:
Bleh nope, I needed this previously but ended up not needing it after refactoring and forgot to remove it. I'll remove it (and make `foreachTensorInplace` do exactly what it was doing before)

zou3519 commented on functorch/csrc/init.cpp:
```diff
@@ -29,10 +30,56 @@ static bool has_level(const Tensor& self, int64_t level) {
 }
 
+static bool has_functional_level(const Tensor& self, int64_t level) {
+  const auto* functional = dynamic_cast<FunctionalTensorWrapper*>(self.unsafeGetTensorImpl());
```
The alternative to using dynamic_cast (if I am remembering how functional tensor works) could be to check the tensor's dispatch key set and see if it has "functionalize" in it

bdhirsh replied:
yep i remember this coming up before - good point, I should just use static cast and check based on whether or not the keyset includes the functionalize key.

zou3519 commented on functorch/csrc/DynamicLayer.cpp:
```diff
@@ -457,6 +462,18 @@ static DispatchKeySet keysToExcludeWhenEnteringDynamicLayer(DispatchKey key) {
   exclude = exclude - keysForEnteringDynamicLayer(key);
   return exclude;
 }
+static bool functionalAtCurrentLevel(const Tensor& tensor) {
```
nit: isFunctionalTensorAtCurrentLevel, "functional" is an overloaded word by itself

zou3519 commented on functorch/csrc/DynamicLayer.cpp:
```diff
@@ -499,6 +516,17 @@ void dynamicLayerFrontFallback(const c10::OperatorHandle& op, torch::jit::Stack*
     }
     hacky_include = hacky_include.add(kVmapModeKey);
+  } else if (layer.key() == DispatchKey::Functionalize) {
+    const auto args = torch::jit::last(stack, op.schema().arguments().size());
+    bool any_tensor_args = anyTensors(args, [&](const Tensor& tensor) { return true; });
+    bool any_functional_args = anyTensors(args, functionalAtCurrentLevel);
+    // Only enable dispatch on Functionalize if either:
+    // - there are any functional tensors at the current level
+    // - we hit a factory op
+    if (!any_functional_args && any_tensor_args) {
+      exclude = exclude.add(DispatchKey::Functionalize);
+    }
+    hacky_include = hacky_include.add(DispatchKey::Functionalize);
```
Hmm, why do we need this?

We have a similar special case above for vmap, but I'm about to get rid of it because it doesn't actually work for composite operations.

zou3519 replied:
If none of the tensors are FunctionalTensors, then having the functional key enabled will mean we just skip directly to DynamicLayerBack, and then come back to DynamicLayerFront, which will check the next thing on the stack. This behavior seems OK to me but I'm probably missing something

bdhirsh replied:
Hmm, I think we need it in order to functionalization to handle factory ops properly. I thought that vmap had the same problem but I might be wrong - how do you ensure that you dispatch to the batching rule properly for factory ops?

> If none of the tensors are FunctionalTensors, then having the functional key enabled will mean we just skip directly to DynamicLayerBack, and then come back to DynamicLayerFront, which will check the next thing on the stack

But what I think I want to happen here is that we hit a `Functionalize` kernel *before* getting to `DynamicLayerBack`, since what that'll do is call the underlying factory op, and then wrap the output tensor into a `FunctionalTensorWrapper`. That's basically why I wanted to keep the functionalize key in the include set here.

zou3519 commented on functorch/_src/eager_transforms.py:
```diff
@@ -1076,3 +1082,47 @@ def wrapper(*args, **kwargs):
+
+def functionalize(func: Callable) -> Callable:
+    @wraps(func)
+    def wrapped(*args, **kwargs):
```
kwargs might have Tensors. So do we just ignore them?

bdhirsh replied:
(I was before, but not for any good reason - when I switch to using tree_map I'll include the kwargs)

zou3519 commented on functorch/_src/eager_transforms.py:
```diff
@@ -1076,3 +1082,47 @@ def wrapper(*args, **kwargs):
+    def wrapped(*args, **kwargs):
+        try:
+            func_level = _func_increment_nesting()
+            func_args = [_wrap_functional_tensor(x, func_level) if isinstance(x, torch.Tensor) else x for x in args]
```
We could just tree_map _wrap_functional_tensor on the Tensors to support cases where the user passes in e.g. dicts of tensors. Otherwise, it might be a good idea to assert that the input is a flat list of tensors.

bdhirsh replied:
Yep thanks - tree_map() seems like the right thing here (I'll probably take a look at how it's used in the other transforms)

zou3519 commented on functorch/_src/eager_transforms.py:
```diff
@@ -1076,3 +1082,47 @@ def wrapper(*args, **kwargs):
+                    torch._sync(a)
+
+            propagate_functional_input_mutations(unwrapped_tensor_inputs, wrapped_tensor_inputs)
+            return [_unwrap_functional_tensor(x) if isinstance(x, torch.Tensor) else x for x in flattened_outputs]
```
If the original function returns a single tensor output, then functionalize(f) is going to return a List, right?

The way to get this to return something of the same python structure as func_outputs is to do a tree_unflatten using the spec obtained from `flattened_outputs, spec = tree_flatten(func_outputs)`

Or you can just `output = tree_map(something_here, func_outputs)` where something_here is `_unwrap_functional_tensor(x) if isinstance(x, torch.Tensor) else x`

zou3519 commented on functorch/_src/eager_transforms.py:
```diff
@@ -1076,3 +1082,47 @@ def wrapper(*args, **kwargs):
+
+            propagate_functional_input_mutations(unwrapped_tensor_inputs, wrapped_tensor_inputs)
+            return [_unwrap_functional_tensor(x) if isinstance(x, torch.Tensor) else x for x in flattened_outputs]
+        finally:
+            _func_decrement_nesting()
+    return wrapped
```
I added some suggestions for UX for this transform. tbh, when we started writing other transforms (e.g. vmap, grad), we just asserted that all inputs were tensors and that all outputs are tensors for simplicity so we didn't need to deal with tree_map. So feel free to ignore the comments above

zou3519 commented on functorch/_src/eager_transforms.py:
```diff
@@ -1076,3 +1082,47 @@ def wrapper(*args, **kwargs):
+        unwrapped_tensor_inputs: List[torch.Tensor],
+        wrapped_tensor_inputs: List[torch.Tensor]) -> None:
+    if len(unwrapped_tensor_inputs) != len(wrapped_tensor_inputs):
+        raise Exception()
```
Probably want to write something here? :P

bdhirsh replied:
this should really be an internal assert (it shouldn't fail from any user inputs), but I'll definitely add something 😛

zou3519 commented on functorch/csrc/init.cpp:
```diff
@@ -29,10 +30,56 @@ static bool has_level(const Tensor& self, int64_t level) {
+  wrapped_impl->regenerate_from_base();
+  auto& wrapped_inner = wrapped_impl->value();
+  // It would probably be more reasonable to check that the two tensors are aliased,
+  // but we can't do that unless we give BatchedTensorImpl a notion of storage.
+  if (unwrapped.unsafeGetTensorImpl() == wrapped_inner.unsafeGetTensorImpl()) {
```
I think a lot of people have been asking for an is_alias_of operator, see https://github.com/pytorch/pytorch/pull/71378

zou3519 commented on functorch/csrc/init.cpp:
```diff
@@ -29,10 +30,56 @@ static bool has_level(const Tensor& self, int64_t level) {
+  } else {
+      TORCH_INTERNAL_ASSERT(unwrapped.nbytes() == wrapped_inner.nbytes());
+      if (unwrapped.sizes() == wrapped_inner.sizes()) {
+          std::cout << "propagating changes to a mutated input!" << std::endl;
```
Should this be here?

bdhirsh replied:
this was also for my debugging sanity, deleting

zou3519 commented on functorch/csrc/init.cpp:
```diff
@@ -29,10 +30,56 @@ static bool has_level(const Tensor& self, int64_t level) {
+          std::cout << "propagating changes to a mutated input!" << std::endl;
+          unwrapped.copy_(wrapped_inner);
+      } else {
+          std::cout << "An inplace-mutation op (like transpose_() was called on an input to the functionalization pass.";
+          std::cout << " Propagating those mutations to the input is currently not supported." << std::endl;
```
Should this be an error message? Or a warning?

bdhirsh replied:
I'm thinking of making it an error, but let me know if you have an opinion. Hopefully we can remove `transpose_()` et al in the near future anyway.

zou3519 commented on functorch/csrc/init.cpp:
```diff
@@ -99,6 +146,14 @@ Tensor _remove_batch_dim(const Tensor& self, int64_t level, int64_t batch_size,
 }
 
+Tensor _unwrap_functional_tensor(const Tensor& self) {
+  auto* functional = dynamic_cast<FunctionalTensorWrapper*>(self.unsafeGetTensorImpl());
+  // We only ever call that after popping out of a functionalize() call, in which case the current tensors
+  // should always be wrapped in a FunctionalTensorWrapper.
+  TORCH_INTERNAL_ASSERT(functional != nullptr);
+  return functional->value();
```
Do we need to _sync here? Or is it already guaranteed that all the queued updates have been applied?

bdhirsh replied:
we do, and this actually fixes a bug. Thanks!

zou3519 commented on functorch/csrc/TensorWrapper.cpp:
```diff
@@ -185,11 +185,29 @@ void dead_tensor_wrapper_fallback(const c10::OperatorHandle& op, torch::jit::Sta
   op.callBoxed(stack);
 }
 
+Tensor& copy_wrapper_tensor_(Tensor& self, const Tensor& src, bool non_blocking) {
+  const auto* self_impl = maybeGetTensorWrapper(self);
+  const auto* src_impl = maybeGetTensorWrapper(src);
+  if (!self_impl || !src_impl || self_impl->level() != src_impl->level()) {
+    TORCH_WARN("The functionalization pass encountered a function that mutated its inputs \
+                and it was unable to preserve the mutations. Inputs to the function \
+                will not be mutated");
+    return self;
+  }
+  return self_impl->value().copy_(src_impl->value(), non_blocking);
+}
+
 // TensorWrapper backend fallback: Unwrap and fallthrough.
 
 TORCH_LIBRARY_IMPL(_, FT_GRAD_WRAPPER_KEY, m) {
   m.fallback(torch::CppFunction::makeFromBoxedFunction<&dead_tensor_wrapper_fallback>());
 }
 
+TORCH_LIBRARY_IMPL(aten, FT_GRAD_WRAPPER_KEY, m) {
+  // In order to preserve semantics for functions that mutate inputs,
+  // an input tensor wrapper needs to know one mutate operation: copy_.
+  m.impl("copy_", copy_wrapper_tensor_);
+}
+
```
Sorry, what is going on here? TensorWrapper (which I should really rename GradTrackingTensor or Variable because that's what it's being used as in functorch) should be unrelated to functionalization, right?

bdhirsh replied:
darn i thought i removed this - that was a relic from an earlier version of the functionalization pass.

zou3519 commented on functorch/csrc/DynamicLayer.cpp:
```diff
@@ -513,6 +541,20 @@ void dynamicLayerFrontFallback(const c10::OperatorHandle& op, torch::jit::Stack*
 
   // Re-dispatch
   op.callBoxed(stack);
+  // checking exclude set because we should only do this if we actually hit a functionalization kernel.
+  if (layer.key() == DispatchKey::Functionalize && !exclude.has(DispatchKey::Functionalize)) {
```
Checking `!exclude.has(DispatchKey::Functionalize)` is a little fishy here, I need to think a bit more about it. I'm worried that in callBoxed if we have nested functionalize transforms then some of them might reset the exclude? But we don't support nested functionalize transforms yet so this is probably fine for now...

zou3519 commented on functorch/csrc/DynamicLayer.cpp:
```diff
@@ -594,6 +637,8 @@ void dynamicLayerBackFallback(const c10::OperatorHandle& op, torch::jit::Stack*
     }
     // Step 2
     foreachTensorInplace(*stack, stack->size() - args_size, stack->size(), unwrap);
+  } else if (cur_key == DispatchKey::Functionalize) {
+    sanityCheckNotFunctional(op, stack, args_size);
```
Makes sense, but we'll need to change this for nested functionalize, right? (if that's a correct assessment then a comment here would be a nice reminder)

bdhirsh replied:
yes, and agreed - i'll add a comment to make it clear that this should be taken out when we eventually have nested functionalization support (I have it there mostly for sanity to help with debugging)

zou3519 commented on functorch/csrc/DynamicLayer.cpp:
```diff
@@ -513,6 +541,20 @@ void dynamicLayerFrontFallback(const c10::OperatorHandle& op, torch::jit::Stack*
+  // checking exclude set because we should only do this if we actually hit a functionalization kernel.
+  if (layer.key() == DispatchKey::Functionalize && !exclude.has(DispatchKey::Functionalize)) {
+    auto ret_size = op.schema().returns().size();
+    foreachTensorInplace(*stack, stack->size() - ret_size, stack->size(),
+      [&](const Tensor& tensor) {
+        TORCH_INTERNAL_ASSERT(at::functionalization::impl::isFunctionalTensor(tensor));
+        auto wrapper = at::functionalization::impl::unsafeGetFunctionalWrapper(tensor);
+        // Functorch is responsible for setting the level on the wrapper, since we don't
+        // have that info available in core (for now).
+        wrapper->set_level(layer.layerId());
+        return tensor;
```
Ugh this is pretty hacky, would it be better to have functional tensor in core propagate its level instead of rely on DispatchKey::Functionalize being in the exclude set?

bdhirsh replied:
Totally agreed - I just can't think of a reasonable way this in core, at least for now. I'm hoping that we can handle this more gracefully if/when the dynamic layering logic makes its way into core.

The reason we can't really teach functional tensors to propagate their levels for now is really just.. factory operators. For non-factory ops, we can just read off the level from the input tensors to set the outputs. But when we hit a factory op, when you're in the functionalization kernel you have no way of knowing what level you are.

zou3519 reviewed (commented):
Need to run, but here are some initial comments and questions

bdhirsh commented:
I think this is ready for another round (pending CI), I finally addressed the comments. Let me know if you think it's worth talking more about the factory op / TLS issue

zou3519 reviewed (approved):
LGTM, left some minor comments in-line.

I'm wondering about the test plan after this PR. The main use case for the functionalize() transform would be to support AOTAutograd. @Chillee, @anijain2305, are you folks planning on testing the transform after this PR?

bdhirsh commented:
(I'll confirm, but the failing tests should be fixed by https://github.com/pytorch/pytorch/pull/73440)
